### PR TITLE
[X86][BF16] Do not combine FP_TRUNC + FP_EXTEND if they come from user

### PR DESCRIPTION
**Pull request by phoebewang**
As discussed in https://github.com/llvm/llvm-project/commit/3cf8535dbf0bf5fafa99ea1f300e2384a7254fba We are not allowed to combine explicit fptrunc/fpext from user.

### CODE REVIEW

File: llvm/test/CodeGen/X86/bfloat.ll [hunk 1]
```diff
@@ -2420,3 +2420,47 @@ define <16 x bfloat> @concat_dup_v8bf16(<8 x bfloat> %x, <8 x bfloat> %y) {
...
+; X86-NEXT:    vmovd %eax, %xmm0
+; X86-NEXT:    vmovd %xmm0, (%esp)
+; X86-NEXT:    flds (%esp)
+; X86-NEXT:    popl %eax
+; X86-NEXT:    retl
+;
+; SSE2-LABEL: trunc_ext:
+; SSE2:       # %bb.0:
+; SSE2-NEXT:    pushq %rax
+; SSE2-NEXT:    callq __truncsfbf2@PLT
```
Comment (rmlarsen) [inline llvm/test/CodeGen/X86/bfloat.ll:2441]: Is there a way to use a rounding conversion here, even if it's slower? Are there tests for this that check that rounding is used?

Comment (phoebewang) [reply llvm/test/CodeGen/X86/bfloat.ll:2441]: What's the rounding conversion mean? A non-default rounding (RNE)? I don't think compiler guarantees rounding correntness except for the default rounding mode in default FP model. Only strict FP model guarantees it.

Comment (krzysz00) [reply llvm/test/CodeGen/X86/bfloat.ll:2441]: __truncsfbf2 is usually defined as RNE in the implementations I've seen of it

**Review by krzysz00 — COMMENTED** [review]
Seems fine to me? But I'm also not competent to review this

File: llvm/test/CodeGen/X86/bfloat.ll (hunk 1 above)
Comment (rmlarsen) [reply llvm/test/CodeGen/X86/bfloat.ll:2441]: If that's the case, then this is fine by me.

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 2]
```diff
@@ -56732,6 +56732,7 @@ static SDValue combineFP16_TO_FP(SDNode *N, SelectionDAG &DAG,
 }
 
 static SDValue combineFP_EXTEND(SDNode *N, SelectionDAG &DAG,
+                                TargetLowering::DAGCombinerInfo &DCI,
```
[the remark this answers is not in the document]
Comment (phoebewang) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:56735]: Legalizer usually only focuses on a single node, while combiner works for multi nodes.

I think we can do it to `half` too. Let's start from BF16 and see if there's any other objections.

**Comment by phoebewang** [comment]
Ping?

**Review by krzysz00 — APPROVED** [review]
LGTM

**Pull request merged**